Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -855,6 +943,321 @@ lemma addY_of_Z_ne_zero {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation
   erw [addY, negY_of_Z_ne_zero <| addZ_ne_zero_of_X_ne hx, addX_of_Z_ne_zero hP hQ hPz hQz hx,
     negAddY_of_Z_ne_zero hP hQ hPz hQz hx, Affine.addY]
 
+variable (W') in
+/-- The coordinates of the addition of two distinct point representatives. -/
+noncomputable def addXYZ (P Q : Fin 3 → R) : Fin 3 → R :=
+  ![W'.addX P Q, W'.addY P Q, addZ P Q]
+
+lemma addXYZ_smul (P Q : Fin 3 → R) (u v : R) :
+    W'.addXYZ (u • P) (v • Q) = (u * v) ^ 2 • W'.addXYZ P Q := by
+  rw [addXYZ, addX_smul, addY_smul, addZ_smul]
+  rfl
+
+lemma addXYZ_of_Z_eq_zero_left' {P Q : Fin 3 → R} (hP : W'.Equation P) (hPz : P z = 0) :
+    W'.addXYZ P Q = ![(P x * Q z) ^ 2 * Q x, (P x * Q z) ^ 3 * Q y, P x * Q z * Q z] := by
+  rw [addXYZ, addX_of_Z_eq_zero_left hPz, addY_of_Z_eq_zero_left hP hPz, addZ_of_Z_eq_zero_left hPz]
+
+lemma addXYZ_of_Z_eq_zero_left {P Q : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z = 0) :
+    W.addXYZ P Q = (P x * Q z) • Q :=
+  addXYZ_of_Z_eq_zero_left' hP.left hPz
```

Comment:
Why is it necessary to have a version with the scalar multiplication expanded out?

Reply:
Because they don't need the assumption that we're working over a field, but I don't know how useful this would be.

Reply:
But this supposedly also works?
```suggestion
lemma addXYZ_of_Z_eq_zero_left {P Q : Fin 3 → R} (hP : W.Equation P) (hPz : P z = 0) :
    W.addXYZ P Q = (P x * Q z) • Q := by
  rw [smul_fin3, addXYZ, addX_of_Z_eq_zero_left hPz, addY_of_Z_eq_zero_left hP hPz, addZ_of_Z_eq_zero_left hPz]
```

Reply:
I guess it does...